docs(dev-style): remove rule about variable declarations

kylo252 opened the pull request:
~~Apply fixes to violations detected by clang-tidy to conform with Neovim's style guide~~

Update style guide for variable declaration as per suggestion in https://github.com/neovim/neovim/pull/20446#issuecomment-1265905231

```help
Multiple, simple variable declarations can be declared on the same line given
that there are no initializations. A simple variable declaration is one that
is not a pointer or array. >
    int i, j, k;    // Allowed
    
Use only one initialization per line. >
    int i, j = 1;   // BAD
```

ref: [DCL04-C. Do not declare more than one variable per declaration](https://wiki.sei.cmu.edu/confluence/display/c/DCL04-C.+Do+not+declare+more+than+one+variable+per+declaration)

justinmk commented:
I am not sure (can't remember) if this style rule is mandatory. I assume that's why it wasn't enforced by uncrustify. If this causes a lot of churn in 100s of files we might want to skip it. What do others think?

zeertzjq commented:
I don't think this is an improvement.

dundargoc commented:
Pretty sure this isn't mentioned in the neovim style guide.

Regardless, we've already discussed this. The conclusion was that separating declarations with initializations is fine since that is what is actually bugprone, and what this rule is trying to guard against. 

Separating empty declarations is useless noise. The correct solution for these is to instead initialize them when they are being declared, rather than just spamming empty declarations.

Since clang-tidy cannot separate these two the idea was scrapped since it created an awful amount of unnecessary change with little gain.

I'd say separate declarations with initialization, and skip empty declarations in this PR.

kylo252 commented:
> Pretty sure this isn't mentioned in the neovim style guide.

It is :smiling_face_with_tear: tho, see "Variable declarations" in https://neovim.io/doc/user/dev_style.html#dev-style 

> Regardless, we've already discussed this. The conclusion was that separating declarations with initializations is fine since that is what is actually bugprone, and what this rule is trying to guard against.

> Separating empty declarations is useless noise. The correct solution for these is to instead initialize them when they are being declared, rather than just spamming empty declarations.

Good points there, but you missed an important factor, from [Google's style-guide](https://google.github.io/styleguide/cppguide.html#Pointer_and_Reference_Expressions)

"_It is allowed (if unusual) to declare multiple variables in the same declaration, but it is disallowed if any of those have pointer or reference decorations. Such declarations are easily misread._"
```c
// Fine if helpful for readability.
int x, y;

int x, *y;  // Disallowed - no & or * in multiple declaration
int* x, *y;  // Disallowed - no & or * in multiple declaration; inconsistent spacing
char * c;  // Bad - spaces on both sides of *
const std::string & str;  // Bad - spaces on both sides of &
```

dundargoc commented:
> see "Variable declarations" in https://neovim.io/doc/user/dev_style.html#dev-style

Well shit. Welp, ignore what I said :).

justinmk commented:
let's change the style rule to:

```
Variable declarations ~

Prefer one variable per line. But this is not mandatory and not enforced by lint/autoformat. >

    int i;
    int j = 1;
```

dundargoc commented:
Just remove the rule. If we don't intend to enforce it it's just noise.

Do you want to pivot this PR and remove it/change to Justin's suggestion? If not can we close this?

justinmk commented:
my suggestion https://github.com/neovim/neovim/pull/20446#issuecomment-1265905231 was to adjust the style guideline in `:help dev-style` and remove any clint rule or other rule enforcing this.

kylo252 committed:
docs(dev): update style guide

kylo252 commented:
> That means we don't need to go back and update existing code like this PR does.

@justinmk, did you mean removing the rule completely?

dundargoc commented:
> I've already did that since last week 6f455ea 😄

What? I'm suggesting you should:

1. Revert the changes you have made.
2. Either remove this rule from the dev-style docs completely or change it like Justin said.

Clint doesn't have this rule, we'd known it by now otherwise.

dundargoc committed:
docs(dev-style): remove rule about variable declarations

The other style rules such as "initialize variables in the declaration"
should already take care of this rule automatically.
